### PR TITLE
Add @Signal Macro

### PR DESCRIPTION
**Pull request by samdeane**
Implements #586.
See also previous discussion on #584.

## TL;DR

This PR adds a new `@Signal` macro, giving a cleaner and more consistent way to add user-defined signals from Swift. 

```swift

@Godot
class MyNode: Node {
  @Signal var mySignal: SimpleSignal
  @Signal var signalWithArgs: SignalWithArguments<Int, String>
}
```

These signals can be used in the same way as built-in signals:

```swift
node.mySignal.connect { /* do something */ }

node.signalWithArgs.connect { intArg, stringArg in
  // do something
}

node.mySignal.emit()
node.signalWithArgs.emit(1, "foo")
```

## Motivation

The existing `#signal` macro defines the signal as a **static** property on the class that contains it.

This is different from the generated implementation of built-in signals, and means that we have to use a different syntax to work with them -- which feels... non-optimal.

## Implementation

We add a new `@Signal` macro expansion which generates a computed property on the class, using `GenericSignal`. 
This is the same approach that is used to implement the signals imported from the Godot API.

We also add an `emit()` method to `GenericSignal`, which is of course essential to support emitting signals defined in Swift.

We add `SimpleSignal` and `SignalWithArguments` as slightly clearer aliases for `GenericSignal`

This change should be 100% additive, and should not break the old `#signal` macro.

## To Do

If this change is adopted, it's probably worth updating the documentation and deprecating the old macro.

### CODE REVIEW

File: Generator/Generator/MethodGen.swift [hunk 1]
```diff
@@ -30,6 +30,10 @@ enum GeneratedMethodKind {
     case utilityFunction
 }
 
+/// The list of static method names to make public.
+/// Most of the names are kept fileprivate.
+let publicMethodNames = ["method_get_class", "method_emit_signal"]
```
Comment (samdeane) [inline Generator/Generator/MethodGen.swift:35]: I had to make this public in order to implement an alternative version of `emitSignal`, which the `GenericSignal` implementation needs in order to implement an `emit()` method with the correct arguments. See also `RawCall.swift`.

I suspect that there may be a better way. Possibly I can use a Variant to make the call?

File: Sources/SwiftGodot/Core/RawCall.swift [hunk 2]
```diff
@@ -0,0 +1,45 @@
+@_implementationOnly import GDExtension
+
+extension Object {
+
+    /// Make a raw call to a method bind.
+    /// All input arguments must be marshaled into `Variant`s.
+    /// The result is a `Variant` that must be unmarshaled into the expected type.
+    @discardableResult /* discardable per discardableList: Object, emit_signal */
+    final func rawCall(_ p_method_bind: GDExtensionMethodBindPtr, arguments: [Variant]) -> Variant? {
```
Comment (samdeane) [inline Sources/SwiftGodot/Core/RawCall.swift:9]: I can't help feeling that there's a better way to do this!
Calls to the API are all generated, with the object handle as the first parameter, and the other arguments as a variadic parameter.

I needed to instead make a call that takes an array of arguments.

**Comment by migueldeicaza** [comment]
This is lovely!   Would you mind testing the Object.callv approach?  If it works, that might be better to maintain in the long run.

**Comment by samdeane** [comment]
Using `Object.callv` seems to work well.
Have also squashed all the commits for a cleaner history.

**Comment by migueldeicaza** [comment]
Thank you!

Merging.
